cloud/gcp: TestFileDoesNotExist failed

cloud/gcp.TestFileDoesNotExist [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_CloudUnitTests/17783777?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_CloudUnitTests/17783777?buildTab=artifacts#/) on master @ [cd7b29726f13528f2c435cd46a9f7af2b341cb1a](https://github.com/cockroachdb/cockroach/commits/cd7b29726f13528f2c435cd46a9f7af2b341cb1a):


```
=== RUN   TestFileDoesNotExist
    gcs_storage_test.go:377: 
        	Error Trace:	pkg/cloud/gcp/gcs_storage_test.go:377
        	Error:      	Should be true
        	Test:       	TestFileDoesNotExist
--- FAIL: TestFileDoesNotExist (0.74s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/disaster-recovery
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestFileDoesNotExist.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-44448

cloud/gcp.TestFileDoesNotExist [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_CloudUnitTests/17789908?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_CloudUnitTests/17789908?buildTab=artifacts#/) on master @ [69fbd56c594624b45f421617728ff00462020ff5](https://github.com/cockroachdb/cockroach/commits/69fbd56c594624b45f421617728ff00462020ff5):


```
=== RUN   TestFileDoesNotExist
    gcs_storage_test.go:377: 
        	Error Trace:	pkg/cloud/gcp/gcs_storage_test.go:377
        	Error:      	Should be true
        	Test:       	TestFileDoesNotExist
--- FAIL: TestFileDoesNotExist (0.70s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #135352 cloud/gcp: TestFileDoesNotExist failed [A-disaster-recovery C-test-failure O-robot T-disaster-recovery branch-release-23.2 release-blocker]
- #135351 cloud/gcp: TestFileDoesNotExist failed [A-disaster-recovery C-test-failure O-robot T-disaster-recovery branch-release-24.2 release-blocker]
- #135350 cloud/gcp: TestFileDoesNotExist failed [A-disaster-recovery C-test-failure O-robot T-disaster-recovery branch-release-24.1 release-blocker]
- #135349 cloud/gcp: TestFileDoesNotExist failed [A-disaster-recovery C-test-failure O-robot T-disaster-recovery branch-release-24.3.0-rc release-blocker]
- #135348 cloud/gcp: TestFileDoesNotExist failed [A-disaster-recovery C-test-failure O-robot T-disaster-recovery branch-release-24.3 release-blocker]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestFileDoesNotExist.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

cloud/gcp.TestFileDoesNotExist [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_CloudUnitTests/17800780?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_CloudUnitTests/17800780?buildTab=artifacts#/) on master @ [a5ab06046e817627cd0b64390f4ede0c4a5d82ef](https://github.com/cockroachdb/cockroach/commits/a5ab06046e817627cd0b64390f4ede0c4a5d82ef):


```
=== RUN   TestFileDoesNotExist
    gcs_storage_test.go:377: 
        	Error Trace:	pkg/cloud/gcp/gcs_storage_test.go:377
        	Error:      	Should be true
        	Test:       	TestFileDoesNotExist
--- FAIL: TestFileDoesNotExist (0.77s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #135505 cloud/gcp: TestFileDoesNotExist failed [A-disaster-recovery C-test-failure O-robot T-disaster-recovery branch-release-23.1 release-blocker]
- #135352 cloud/gcp: TestFileDoesNotExist failed [A-disaster-recovery C-test-failure O-robot T-disaster-recovery branch-release-23.2 release-blocker]
- #135351 cloud/gcp: TestFileDoesNotExist failed [A-disaster-recovery C-test-failure O-robot T-disaster-recovery branch-release-24.2 release-blocker]
- #135350 cloud/gcp: TestFileDoesNotExist failed [A-disaster-recovery C-test-failure O-robot T-disaster-recovery branch-release-24.1 release-blocker]
- #135349 cloud/gcp: TestFileDoesNotExist failed [A-disaster-recovery C-test-failure O-robot T-disaster-recovery branch-release-24.3.0-rc release-blocker]
- #135348 cloud/gcp: TestFileDoesNotExist failed [A-disaster-recovery C-test-failure O-robot T-disaster-recovery branch-release-24.3 release-blocker]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestFileDoesNotExist.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

cloud/gcp.TestFileDoesNotExist [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_CloudUnitTests/17810009?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_CloudUnitTests/17810009?buildTab=artifacts#/) on master @ [864b6c3c540a7fb2039227d25e3d785cbdbc769f](https://github.com/cockroachdb/cockroach/commits/864b6c3c540a7fb2039227d25e3d785cbdbc769f):


```
=== RUN   TestFileDoesNotExist
    gcs_storage_test.go:377: 
        	Error Trace:	pkg/cloud/gcp/gcs_storage_test.go:377
        	Error:      	Should be true
        	Test:       	TestFileDoesNotExist
--- FAIL: TestFileDoesNotExist (0.68s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #135505 cloud/gcp: TestFileDoesNotExist failed [A-disaster-recovery C-test-failure O-robot T-disaster-recovery branch-release-23.1 release-blocker]
- #135352 cloud/gcp: TestFileDoesNotExist failed [A-disaster-recovery C-test-failure O-robot T-disaster-recovery branch-release-23.2 release-blocker]
- #135351 cloud/gcp: TestFileDoesNotExist failed [A-disaster-recovery C-test-failure O-robot T-disaster-recovery branch-release-24.2 release-blocker]
- #135350 cloud/gcp: TestFileDoesNotExist failed [A-disaster-recovery C-test-failure O-robot T-disaster-recovery branch-release-24.1 release-blocker]
- #135349 cloud/gcp: TestFileDoesNotExist failed [A-disaster-recovery C-test-failure O-robot T-disaster-recovery branch-release-24.3.0-rc release-blocker]
- #135348 cloud/gcp: TestFileDoesNotExist failed [A-disaster-recovery C-test-failure O-robot T-disaster-recovery branch-release-24.3 release-blocker]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestFileDoesNotExist.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/cloud/external_storage.go
+++ b/pkg/cloud/external_storage.go
@@ -128,6 +128,10 @@ type SQLConnI interface {
 // ErrFileDoesNotExist is a sentinel error for indicating that a specified
 // bucket/object/key/file (depending on storage terminology) does not exist.
 // This error is raised by the ReadFile method.
+//
+// On GCS, a non-existent bucket returns a 403 forbidden, so a not found is
+// only returned if the bucket exists and the user has the permissions needed
+// to interact with the bucket.
 var ErrFileDoesNotExist = errors.New("external_storage: file doesn't exist")
 
 // ErrListingUnsupported is a marker for indicating listing is unsupported.

--- a/pkg/cloud/gcp/gcs_storage_test.go
+++ b/pkg/cloud/gcp/gcs_storage_test.go
@@ -25,7 +25,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
 	"github.com/cockroachdb/cockroach/pkg/util/ioctx"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
-	"github.com/cockroachdb/errors"
 	"github.com/stretchr/testify/require"
 	"golang.org/x/oauth2/google"
 )
@@ -327,8 +326,8 @@ func TestAntagonisticGCSRead(t *testing.T) {
 	cloudtestutils.CheckAntagonisticRead(t, conf, testSettings)
 }
 
-// TestFileDoesNotExist ensures that the ReadFile method of google cloud storage
-// returns a sentinel error when the `Bucket` or `Object` being read do not
+// TestFileDoesNotExist ensures that the ReadFile method of google cloud
+// storage returns a sentinel error when the `Object` being read does not
 // exist.
 func TestFileDoesNotExist(t *testing.T) {
 	defer leaktest.AfterTest(t)()
@@ -341,41 +340,20 @@ func TestFileDoesNotExist(t *testing.T) {
 
 	testSettings := cluster.MakeTestingClusterSettings()
 
-	{
-		// Invalid gsFile.
-		gsFile := "gs://cockroach-fixtures-us-east1/tpch-csv/sf-1/invalid_region.tbl?AUTH=implicit"
-		conf, err := cloud.ExternalStorageConfFromURI(gsFile, user)
-		require.NoError(t, err)
+	// Invalid gsFile.
+	gsFile := "gs://cockroach-fixtures-us-east1/tpch-csv/sf-1/invalid_region.tbl?AUTH=implicit"
+	conf, err := cloud.ExternalStorageConfFromURI(gsFile, user)
+	require.NoError(t, err)
 
-		s, err := cloud.MakeExternalStorage(context.Background(), conf, base.ExternalIODirConfig{}, testSettings,
-			nil, /* blobClientFactory */
-			nil, /* db */
-			nil, /* limiters */
-			cloud.NilMetrics,
-		)
-		require.NoError(t, err)
-		_, _, err = s.ReadFile(context.Background(), "", cloud.ReadOptions{NoFileSize: true})
-		require.Error(t, err, "")
-		require.True(t, errors.Is(err, cloud.ErrFileDoesNotExist))
-	}
-
-	{
-		// Invalid gsBucket.
-		gsFile := "gs://cockroach-fixtures-us-east1-invalid/tpch-csv/sf-1/region.tbl?AUTH=implicit"
-		conf, err := cloud.ExternalStorageConfFromURI(gsFile, user)
-		require.NoError(t, err)
-
-		s, err := cloud.MakeExternalStorage(context.Background(), conf, base.ExternalIODirConfig{}, testSettings,
-			nil, /* blobClientFactory */
-			nil, /* db */
-			nil, /* limiters */
-			cloud.NilMetrics,
-		)
-		require.NoError(t, err)
-		_, _, err = s.ReadFile(context.Background(), "", cloud.ReadOptions{NoFileSize: true})
-		require.Error(t, err, "")
-		require.True(t, errors.Is(err, cloud.ErrFileDoesNotExist))
-	}
+	s, err := cloud.MakeExternalStorage(context.Background(), conf, base.ExternalIODirConfig{}, testSettings,
+		nil, /* blobClientFactory */
+		nil, /* db */
+		nil, /* limiters */
+		cloud.NilMetrics,
+	)
+	require.NoError(t, err)
+	_, _, err = s.ReadFile(context.Background(), "", cloud.ReadOptions{NoFileSize: true})
+	require.ErrorIs(t, err, cloud.ErrFileDoesNotExist)
 }
 
 func TestCompressedGCS(t *testing.T) {

